File: MegawattVB/MegawattVB/Form1.vb
Imports System.Text.RegularExpressions

Public Class Form1
    Const BASE As Double = 2000
    Const MULTI As Double = 300
    Const EXPRESS_RATE As Double = 1.05

    Private Sub btnClose_Click(sender As Object, e As EventArgs) Handles btnClose.Click
        Me.Close()
    End Sub

    Private Sub btnClear_Click(sender As Object, e As EventArgs) Handles btnClear.Click
        txtDeposit.Clear()
        txtFirst.Clear()
        txtLast.Clear()
        txtNum.Clear()
        txtPhone.Clear()
        lblAdditional.Text = ""
        lblBalance.Text = ""
        lblBase.Text = ""
        lblDeposit.Text = ""
        lblTotal.Text = ""
        chkExpress.Checked = False
    End Sub

    Private Sub btnCalc_Click(sender As Object, e As EventArgs) Handles btnCalc.Click
        'declare local variables
        Dim numPanels As Integer
        Dim firstName As String
        Dim lastName As String
        Dim phoneNo As String
        Dim deposit As Double
        Dim express As Boolean = chkExpress.Checked
        Dim flag As Boolean = True
        If Check_First(txtFirst.Text) Then
            firstName = txtFirst.Text
        Else
            MessageBox.Show("Invalid Entry")
            Return
        End If
        If Check_Last(txtLast.Text) Then
            lastName = txtLast.Text
        Else
            MessageBox.Show("Invalid Entry")
            Return
        End If
        If Check_Phone(txtPhone.Text) Then
            phoneNo = txtPhone.Text
        Else
            MessageBox.Show("Invalid Entry")
            Return
        End If
        If Check_Num(txtNum.Text) Then
            numPanels = Convert.ToInt32(txtNum.Text)
        Else
            MessageBox.Show("Invalid Entry")
            Return
        End If
        If Check_Deposit(txtDeposit.Text) Then
            deposit = Convert.ToDouble(txtDeposit.Text)
        Else
            MessageBox.Show("Invalid Entry")
            Return
        End If

        lblBase.Text = BASE.ToString("c")
        lblAdditional.Text = Calc_Add(numPanels).ToString("c")
        lblTotal.Text = Calc_Total(numPanels, express).ToString("c")
        lblDeposit.Text = deposit.ToString("c")

        'print appropriate value depending on deposit
        If deposit >= Calc_Total(numPanels, express) Then
            lblChanger.Text = "Refund:"
            lblBalance.Text = -Calc_Deposit(deposit, numPanels, express).ToString("c")
        Else
            lblBalance.Text = Calc_Deposit(deposit, numPanels, express).ToString("c")
        End If

    End Sub

    'calculate additional
    Function Calc_Add(ByVal n As Integer) As Double
        Dim finVal = (n - 2) * MULTI
        Return finVal
    End Function

    'calculate deposit or refund
    Function Calc_Deposit(ByVal d As Double, ByVal n As Integer, ByVal b As Boolean) As Double
        Dim finVal = Calc_Total(n, b) - d
        Return finVal
    End Function

    'calculate total cost
    Function Calc_Total(ByVal n As Integer, ByVal b As Boolean) As Double
        Dim finVal = BASE + Calc_Add(n)

        If b Then
            finVal *= EXPRESS_RATE
        End If

        Return finVal
    End Function
    'check valid first name
    Function Check_First(ByVal s As String) As Boolean
        Dim finVal = Regex.IsMatch(s, "^[A-Za-z]+$")

        If Not finVal Then
            txtFirst.Text = "Enter Name"
        End If

        Return finVal
    End Function

    'check valid last name
    Function Check_Last(ByVal s As String) As Boolean
        Dim finVal = Regex.IsMatch(s, "^[A-Za-z]+$")

        If Not finVal Then
            txtLast.Text = "Enter Name"
        End If

        Return finVal
    End Function

    'check valid phone
    Function Check_Phone(ByVal s As String) As Boolean
        Dim finVal = Regex.IsMatch(s, "^(\+\d{1,2}\s)?\(?\d{3}\)?[\s.-]\d{3}[\s.-]\d{4}$")

        If Not finVal Then
            txtPhone.Text = "Enter Phone"
        End If

        Return finVal
    End Function

    'check valid integer
    Function Check_Num(ByVal s As String) As Boolean
        Dim finVal = Regex.IsMatch(s, "^[0-9]+$")

        If Not finVal Then
            txtNum.Text = "Enter Integer"
        Else
            Dim temp As Integer = Convert.ToInt32(s)
            If Not (temp <= 1000 And temp > 0) Then
                finVal = False
            End If
        End If

        Return finVal
    End Function

    'check valid decimal
    Function Check_Deposit(ByVal s As String) As Boolean
        Dim finVal = Regex.IsMatch(s, "^[0-9]*(\.[0-9]{1,2})?$")

        If Not finVal Then
            txtDeposit.Text = "Enter Decimal"
        End If

        Return finVal
    End Function

End Class

File: MegawattC/MegawattC/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;
using System.Threading.Tasks;
using System.Windows.Forms;

namespace MegawattC
{
    public partial class Form1 : Form
    {
        private String first;
        private String last;
        private String phone;
        private int num;
        private double deposit;
        private const double BASE = 2000;
        private const double EXPRESS = .05;
        private const double MULTI = 300;
        public Form1()
        {
            InitializeComponent();
        }
        
        private void btnCalc_Click(object sender, EventArgs e)
        {
            //check all necessary values
            if (Check_First(txtFirst.Text))
                first = txtFirst.Text;
            if (Check_Last(txtLast.Text))
                last = txtLast.Text;
            if (Check_Phone(txtPhone.Text))
                phone = txtPhone.Text;
            if (Check_Num(txtNumPanels.Text))
                num = Convert.ToInt32(txtNumPanels.Text);
            if (Check_Deposit(txtDeposit.Text))
                deposit = Convert.ToDouble(txtDeposit.Text);
            
            if(!(Check_First(txtFirst.Text) && Check_Last(txtLast.Text) && Check_Phone(txtPhone.Text) && Check_Num(txtNumPanels.Text) && Check_Deposit(txtDeposit.Text)))
            {
                MessageBox.Show("Invalid input");
                return;
            }

            //output necessary info
            lblBase.Text = BASE.ToString("c");
            lblAdditional.Text = Calc_Additional(num).ToString("c");
            lblTotal.Text = Calc_Total(num).ToString("c");
            lblDeposit.Text = deposit.ToString("c");
            
            //change balance to refund if necessary
            if(Calc_Total(num) <= deposit)
            {
                lblChanger.Text = "Refund:";
                lblBalance.Text = (-(deposit - Calc_Total(num))).ToString("c");
            }
            else
            {
                lblChanger.Text = "Balance due:";
                lblBalance.Text = (Calc_Total(num) - deposit).ToString("c");
            }
        }
        private double Calc_Total(int num)
        {
            if (chkExpress.Checked)
            {
                return 1.05 * (BASE + Calc_Additional(num));
            }
            else
            {
                return BASE + Calc_Additional(num);
            }
        }
        private double Calc_Additional(int num)
        {
            if (num >= 2)
            {
                return num * MULTI;
            }
            else
            {
                return 0;
            }
        }
        private bool Check_First(String s)
        {
            bool validated = Regex.IsMatch(s, @"^[a-zA-Z]+$");

            if (!validated)
                txtFirst.Text = "Enter Name";

            return validated;
        }
        private bool Check_Last(String s)
        {
            bool validated = Regex.IsMatch(s, @"^[a-zA-Z]+$");

            if (!validated)
                txtLast.Text = "Enter Name";

            return validated;
        }
        private bool Check_Phone(String s)
        {
            bool validated = Regex.IsMatch(s, @"\(?\d{3}\)?-? *\d{3}-? *-?\d{4}");

            if (!validated)
                txtPhone.Text = "Enter Phone";

            return validated;
        }
        private bool Check_Num(String s)
        {
            bool validated = Regex.IsMatch(s, @"^[0-9]+$");

            if (!validated)
                txtNumPanels.Text = "Enter Integer";
            else
            {
                int temp = Convert.ToInt32(s);
                if (!(temp <= 1000 && temp > 0))
                {
                    validated = false;
                }
            }

            return validated;
        }
        private bool Check_Deposit(String s)
        {
            bool validated = Regex.IsMatch(s, @"^[0-9]*(\.[0-9]{1,2})?$");

            if (!validated)
                txtDeposit.Text = "Enter Decimal";

            return validated;
        }

        private void btnClear_Click(object sender, EventArgs e)
        {
            txtFirst.Text = "";
            txtLast.Text = "";
            txtDeposit.Text = "";
            txtNumPanels.Text = "";
            txtPhone.Text = "";
            chkExpress.Checked = false;
            lblAdditional.Text = "";
            lblBalance.Text = "";
            lblBase.Text = "";
            lblDeposit.Text = "";
            lblTotal.Text = "";
        }

        private void btnClose_Click(object sender, EventArgs e)
        {
            this.Close();
        }
    }
}
